test/randomized/test_transpiler_equivalence.py

```diff
@@ -0,0 +1,225 @@
+
+
+class QCircuitMachine(RuleBasedStateMachine):
+    """Build a Hypothesis rule based state machine for constructing, transpiling
+    and simulating a series of random QuantumCircuits.
+
+    Build circuits with up to QISKIT_RANDOM_QUBITS qubits, apply a random
+    selection of gates from qiskit.extensions.standard with randomly selected
+    qargs, cargs, and parameters. At random intervals, transpile the circuit for
+    a random backend with a random optimization level and simulate both the
+    initial and the transpiled circuits to verify that their counts are the
+    same.
+
+    """
+
+    qubits = Bundle('qubits')
+    clbits = Bundle('clbits')
+
+    max_qubits = int(os.getenv('QISKIT_RANDOM_QUBITS', '5'))
+    backends = []
+    for backend in mock_backends:
+        if backend.configuration().n_qubits >= max_qubits:
+            backends.append(backend)
+
+    def __init__(self):
+        super().__init__()
+        self.qc = QuantumCircuit()
+
+    @precondition(lambda self: len(self.qc.qubits) < self.max_qubits)
+    @rule(target=qubits,
+          n=st.integers(min_value=1, max_value=max_qubits))
+    def add_qreg(self, n):
+        """Adds a new variable sized qreg to the circuit, up to max_qubits."""
+        n = min(n, self.max_qubits - len(self.qc.qubits))
+        qreg = QuantumRegister(n)
+        self.qc.add_register(qreg)
+        return multiple(*list(qreg))
+
+    @rule(target=clbits,
+          n=st.integers(1, 5))
+    def add_creg(self, n):
+        """Add a new variable sized creg to the circuit."""
+        creg = ClassicalRegister(n)
+        self.qc.add_register(creg)
+        return multiple(*list(creg))
+
+    # Gates of various shapes
+
+    @rule(gate=st.sampled_from(oneQ_gates),
+          qarg=qubits)
+    def add_1q_gate(self, gate, qarg):
+        """Append a random 1q gate on a random qubit."""
+        self.qc.append(gate(), [qarg], [])
+
+    @rule(gate=st.sampled_from(twoQ_gates),
+          qargs=st.lists(qubits, max_size=2, min_size=2, unique=True))
+    def add_2q_gate(self, gate, qargs):
+        """Append a random 2q gate across two random qubits."""
+        self.qc.append(gate(), qargs)
+
+    @rule(gate=st.sampled_from(threeQ_gates),
+          qargs=st.lists(qubits, max_size=3, min_size=3, unique=True))
+    def add_3q_gate(self, gate, qargs):
+        """Append a random 3q gate across three random qubits."""
+        self.qc.append(gate(), qargs)
+
+    @rule(gate=st.sampled_from(oneQ_oneP_gates),
+          qarg=qubits,
+          param=st.floats(allow_nan=False, allow_infinity=False))
+    def add_1q1p_gate(self, gate, qarg, param):
+        """Append a random 1q gate with 1 random float parameter."""
+        self.qc.append(gate(param), [qarg])
+
+    @rule(gate=st.sampled_from(oneQ_twoP_gates),
+          qarg=qubits,
+          params=st.lists(
+              st.floats(allow_nan=False, allow_infinity=False),
+              min_size=2, max_size=2))
```

Comment:
We can, but I'm not sure it would be beneficial. Values outside `[0,2*pi]` are still valid inputs so we'd want to catch e.g. a bug handling a param of `-pi/2` or `5*pi/2`. If the goal is to constrain the search space to get the test to spend more time on more likely cases (so say, constrain to `[-8*pi,8*pi]`), I'm on the fence. I'd like to believe that hypothesis is aware that most of the commonly used floats will center around zero and so will wisely balance its time between the common and the obscure, but I don't know for sure.

Just to give a feel for what `floats` will generate:
```
>>> [ st.floats(allow_nan=False, allow_infinity=False).example() for _ in range(100) ]
[1e-05, 1.7976931348623157e+308, -3.402823466e+38, 2.551078027060389e+54, -3.402823466e+38, -1.192092896e-07, 8.261929414679991e-129, 1.192092896e-07, -0.3333333333333333, 10000000.0, -1.7976931348623157e+308, 2.00001, -9007199254740992.0, 3.0625586463600532e+16, -1e-05, -2.568602351828513e+291, 10000000.0, 1.192092896e-07, -2.036219043860565e+16, 2.2250738585072014e-308, 6.297658820997385e+16, -10000000.0, -2.220446049250313e-16, -1.5, 1.9, 1.1, -1.1, -10000000.0, 0.3333333333333333, -2.00001, 1.5, 10000000.0, 4.323387237280064e+16, -10000000.0, -6.11961240738528e+16, 1.5, -2.193274026383663e+16, 9.35195918496845e-45, 6.198177208980085e+16, -3.7110798832169737e+74, 2.00001, 0.99999, 1.9, 1.7976931348623157e+308, 0.0, -0.99999, -10000000.0, -1.7976931348623157e+308, 1.192092896e-07, 1.9, 1.5, 2.2250738585072014e-308, -1.7976931348623157e+308, -2.220446049250313e-16, 0.99999, 1.2951731060212834e-06, -3.402823466e+38, -2.00001, -4564596093695066.0, 9007199254740992.0, -1e-05, 3.9148913896605384e+16, 9007199254740992.0, -1e-05, -2.2250738585072014e-308, 1.9, 2.00001, -10000000.0, 2.00001, -3.402823466e+38, -2.220446049250313e-16, -10000000.0, 1.7976931348623157e+308, 0.99999, -1.9, 0.99999, -1e-05, 1.175494351e-38, -2.220446049250313e-16, -1e-05, 2.220446049250313e-16, -1.1, -1.1, -2.2250738585072014e-308, 4.988769545527318e+16, -1e-05, 2.00001, -1.1, -1.5, -2.00001, -1e-05, 1.7976931348623157e+308, -0.5, -1e-05, 1e-05, 0.99999, -10000000.0, -1.175494351e-38, -1.175494351e-38, 1.175494351e-38]
```

```
>>> [ st.floats(allow_nan=False, allow_infinity=False, min_value=0, max_value=2*np.pi).example() for _ in range(100) ]
[2.495498257796553, 4.102156650791777, 4.865456404222431, 0.44412804897051905, 5.513710576881485, 3.6289073250694828, 6.093750340788063, 1.8976588767509177, 4.297677022912716, 3.1795946617716897, 0.2581208921704018, 1.2984842273053538, 6.137715146430749, 4.393547288226175, 1.0758733665577793, 3.842174121974989, 6.115692908705578, 1.9809439273694815, 2.3376204605191186, 1.9319838672148102, 0.056044522684440254, 6.1357368897023585, 1.8888422412087982, 4.537055493980659, 0.8741587085624442, 4.316032532527693, 2.1976089073540694, 0.9877661469062294, 0.32313157773720463, 3.7597932717133817, 1.9501787905881314, 0.14510305208482555, 5.008457195432693, 0.8594106292219699, 5.680020421396735, 3.3676329050789144, 2.045899746337321, 5.534186841996694, 1.2501792058912014, 2.061913716650754, 2.8897803909477613, 3.7961702709867735, 2.9544092353741473, 2.1092525850580857, 4.4715063929721115, 2.8100801549537566, 1.599559490056448, 4.13915318772442, 0.17303056290786695, 2.6202256790395713, 0.44701132524079445, 1.9042205618018944, 3.029130402740923, 3.4693733189680867, 5.896648528943915, 6.068968969732009, 1.5503295028908102, 2.195292844370835, 0.21895223283869153, 3.874519266154521, 3.408999173983722, 2.6431433564659836, 0.906638009363137, 5.980461851285671, 6.241139044949207, 2.089037587688607, 2.7890164985724857, 5.107716640016839, 6.248747129707831, 5.8145530027424694, 1.5423223227002205, 3.6462521739281435, 4.730085417613045, 2.738328005502674, 3.057059936256841, 0.05284052279368682, 1.9430163441180066, 5.988085258274451, 0.19491288279358854, 4.184469081023768, 3.4253846947499547, 4.791454932452348, 0.1244527919008161, 0.39099130333514803, 0.08325062546713471, 5.12736396021666, 5.4133889324029285, 1.7820908124009096, 3.585855800845377, 3.1863138439883034, 2.472272980104531, 5.802619194326929, 3.8892720239509417, 0.07973890355496087, 0.6202032111755001, 0.9257458021294165, 1.887999518668018, 0.35157139696460005, 3.8363576946749514, 4.461040875598085]
```

Reply:
One data point, the following fails with a huge lambda, but passes with one closer to unity:

```
>>> qc = qk.QuantumCircuit(2,1)
>>> qc.cx(1,0)
>>> qc.u2(1.0, 9007199254740992.0, 1)
>>> qc.measure(1,0)
>>> qk.transpile(qc, optimization_level=3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/compiler/transpile.py", line 147, in transpile
    circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_configs)))
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/tools/parallel.py", line 100, in parallel_map
    return [task(values[0], *task_args, **task_kwargs)]
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/compiler/transpile.py", line 168, in _transpile_circuit
    return transpile_circuit(circuit, transpile_config)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/transpiler/transpile_circuit.py", line 62, in transpile_circuit
    return pass_manager.run(circuit)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/transpiler/passmanager.py", line 147, in run
    dag = self._do_pass(pass_, dag, passset.options)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/transpiler/passmanager.py", line 180, in _do_pass
    new_dag = pass_.run(dag)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/transpiler/passes/consolidate_blocks.py", line 78, in run
    unitary = UnitaryGate(Operator(subcirc))  # simulates the circuit
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/extensions/unitary.py", line 58, in __init__
    raise ExtensionError("Input matrix is not unitary.")
qiskit.extensions.exceptions.ExtensionError: 'Input matrix is not unitary.'
```

While this is a bug, I'm not sure it's a high value bug.

Reply:
Ok, good point. Since we don't explicitly disallow floats outside [0, 2*pi], they should probably be part of testing as well.